File: MvvmSample/app/src/main/java/com/abben/mvvmsample/ui/movies/MoviesRepository.java
package com.abben.mvvmsample.ui.movies;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.abben.mvvmsample.bean.Movie;
import com.abben.mvvmsample.common.NetworkBoundResource;
import com.abben.mvvmsample.db.MoviesDataBase;
import com.abben.mvvmsample.network.ApiManager;
import com.abben.mvvmsample.network.ApiResponse;
import com.abben.mvvmsample.util.RateLimiter;
import com.abben.mvvmsample.vo.Resource;
import com.abben.mvvmsample.vo.TypeMovies;

import java.util.List;
import java.util.concurrent.TimeUnit;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

/**
 * Created by abben on 2017/11/20.
 */
public class MoviesRepository {

    private final MutableLiveData<Boolean> mMoviesRefreshing;

    private RateLimiter<String> moviesListRateLimiter = new RateLimiter<>(5, TimeUnit.MINUTES);
    private MoviesDataBase moviesDataBase;

    public MoviesRepository(MoviesDataBase moviesDataBase) {
        mMoviesRefreshing = new MutableLiveData<>();
        this.moviesDataBase = moviesDataBase;
    }

    public LiveData<Resource<List<Movie>>> getMovies(final String type) {
        mMoviesRefreshing.setValue(true);
        return new NetworkBoundResource<List<Movie>, List<Movie>>() {

            @Override
            protected void saveCallResult(@NonNull List<Movie> item) {
                moviesDataBase.beginTransaction();
                try {
                    moviesDataBase.moviesDao().insertMovies(item);
                    moviesDataBase.setTransactionSuccessful();
                } finally {
                    moviesDataBase.endTransaction();
                }
            }

            @Override
            protected boolean shouldFetch(@Nullable List<Movie> data) {
                return (data == null || data.isEmpty() || moviesListRateLimiter.shouldFetch(type));
            }

            @NonNull
            @Override
            protected LiveData<List<Movie>> loadFromDb() {
                if (type.equals(TypeMovies.TYPE_ALL_MOVIES_ZH)) {
                    return moviesDataBase.moviesDao().loadAllMovies();
                } else {
                    return moviesDataBase.moviesDao().loadMovies(type);
                }
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<Movie>>> createCall() {
                String apiType;
                switch (type) {
                    case TypeMovies.TYPE_ALL_MOVIES_ZH:
                        apiType = TypeMovies.API_TYPE_ALL_MOVIES;
                        break;
                    case TypeMovies.TYPE_CHINESE_MOVIES_ZH:
                        apiType = TypeMovies.API_TYPE_CHINESE_MOVIES;
                        break;
                    case TypeMovies.TYPE_EURAMERICAN_MOVIES_ZH:
                        apiType = TypeMovies.API_TYPE_EURAMERICAN_MOVIES;
                        break;
                    case TypeMovies.TYPE_JANPAN_AND_KOREA_MOVIES_ZH:
                        apiType = TypeMovies.API_TYPE_JANPAN_AND_KOREA_MOVIES;
                        break;
                    default:
                        apiType = TypeMovies.API_TYPE_ALL_MOVIES;
                        break;
                }
                return ApiManager.getMoviesApi().fetchMovies(apiType);
            }

            @Override
            protected void onFetchFailed() {
                mMoviesRefreshing.setValue(false);
                moviesListRateLimiter.reset(type);
            }

            @Override
            protected void onFetchSuccess() {
                mMoviesRefreshing.setValue(false);
            }
        }.asLiveData();
    }

    public MutableLiveData<Boolean> getRefreshing() {
        return mMoviesRefreshing;
    }
}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/ui/movies/MoviesRepository.kt
package com.abben.mvvmsamplejetpack.ui.movies

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.abben.mvvmsamplejetpack.bean.Movie
import com.abben.mvvmsamplejetpack.common.NetworkBoundResource
import com.abben.mvvmsamplejetpack.db.MoviesDataBase
import com.abben.mvvmsamplejetpack.network.ApiManager
import com.abben.mvvmsamplejetpack.network.ApiResponse
import com.abben.mvvmsamplejetpack.util.RateLimiter
import com.abben.mvvmsamplejetpack.vo.*
import java.util.concurrent.TimeUnit

/**
 *  @author abben
 *  on 2019/1/21
 */
class MoviesRepository(var moviesDataBase: MoviesDataBase) {
    val mMoviesRefreshing: MutableLiveData<Boolean> = MutableLiveData()

    private var moviesListRateLimiter: RateLimiter<String> = RateLimiter(5, TimeUnit.MINUTES)

    fun getMovies(type: String): LiveData<Resource<List<Movie>>> {
        mMoviesRefreshing.value = true
        return object : NetworkBoundResource<List<Movie>, List<Movie>>() {
            override fun saveCallResult(item: List<Movie>) {
                moviesDataBase.beginTransaction()
                try {
                    moviesDataBase.moviesDao().insertMovies(item)
                    moviesDataBase.setTransactionSuccessful()
                } finally {
                    moviesDataBase.endTransaction()
                }
            }

            override fun shouldFetch(data: List<Movie>): Boolean {
                return (data.isEmpty() || moviesListRateLimiter.shouldFetch(type))
            }

            override fun loadFromDb(): LiveData<List<Movie>> {
                return if (type === TYPE_ALL_MOVIES_ZH) {
                    moviesDataBase.moviesDao().loadAllMovies()
                } else {
                    moviesDataBase.moviesDao().loadMovies(type)
                }
            }

            override fun createCall(): LiveData<ApiResponse<List<Movie>>> {
                val apiType = when (type) {
                    TYPE_ALL_MOVIES_ZH -> TypeMovies.API_TYPE_ALL_MOVIES
                    TYPE_CHINESE_MOVIES_ZH -> TypeMovies.API_TYPE_CHINESE_MOVIES
                    TYPE_EURAMERICAN_MOVIES_ZH -> TypeMovies.API_TYPE_EURAMERICAN_MOVIES
                    TYPE_JANPAN_AND_KOREA_MOVIES_ZH -> TypeMovies.API_TYPE_JANPAN_AND_KOREA_MOVIES
                    else -> TypeMovies.API_TYPE_ALL_MOVIES
                }
                return ApiManager.getMoviesApi().fetchMovies(apiType)
            }

            override fun onFetchFailed() {
                mMoviesRefreshing.value = false
                moviesListRateLimiter.reset(type)
            }

            override fun onFetchSuccess() {
                mMoviesRefreshing.value = false
            }
        }.asLiveData()
    }

}